crates/ruff_python_formatter/src/statement/clause.rs

```diff
@@ -0,0 +1,350 @@
+use ruff_python_ast::{
+    ElifElseClause, ExceptHandlerExceptHandler, MatchCase, Ranged, StmtClassDef, StmtFor,
+    StmtFunctionDef, StmtIf, StmtMatch, StmtTry, StmtWhile, StmtWith,
+};
+use ruff_python_trivia::{SimpleToken, SimpleTokenKind, SimpleTokenizer};
+use ruff_text_size::{TextRange, TextSize};
+
+/// The header of a compound statement clause.
+///
+/// > A compound statement consists of one or more ‘clauses.’ A clause consists of a header and a ‘suite.’
+/// > The clause headers of a particular compound statement are all at the same indentation level.
+/// > Each clause header begins with a uniquely identifying keyword and ends with a colon.
+/// [source](https://docs.python.org/3/reference/compound_stmts.html#compound-statements)
+#[derive(Copy, Clone)]
+pub(crate) enum ClauseHeader<'a> {
+    Class(&'a StmtClassDef),
+    Function(&'a StmtFunctionDef),
+    If(&'a StmtIf),
+    ElifElse(&'a ElifElseClause),
+    Try(&'a StmtTry),
+    ExceptHandler(&'a ExceptHandlerExceptHandler),
+    TryFinally(&'a StmtTry),
+    Match(&'a StmtMatch),
+    MatchCase(&'a MatchCase),
+    For(&'a StmtFor),
+    While(&'a StmtWhile),
+    With(&'a StmtWith),
+    OrElse(ElseClause<'a>),
+}
+
+impl<'a> ClauseHeader<'a> {
+    /// The range from the clause keyword up to and including the final colon.
+    pub(crate) fn range(self, source: &str) -> FormatResult<TextRange> {
+        let keyword_range = self.keyword_range(source)?;
+
+        let mut last_child_end = None;
+
+        self.visit(&mut |child| last_child_end = Some(child.end()));
+
+        let end = match self {
+            ClauseHeader::Class(class) => Some(last_child_end.unwrap_or(class.name.end())),
+            ClauseHeader::Function(function) => Some(last_child_end.unwrap_or(function.name.end())),
+            ClauseHeader::ElifElse(_)
+            | ClauseHeader::Try(_)
+            | ClauseHeader::If(_)
+            | ClauseHeader::TryFinally(_)
+            | ClauseHeader::Match(_)
+            | ClauseHeader::MatchCase(_)
+            | ClauseHeader::For(_)
+            | ClauseHeader::While(_)
+            | ClauseHeader::With(_)
+            | ClauseHeader::OrElse(_) => last_child_end,
+
+            ClauseHeader::ExceptHandler(handler) => handler
+                .name
+                .as_ref()
+                .map(ruff_python_ast::Ranged::end)
+                .or(last_child_end),
+        };
+
+        let colon = colon_range(end.unwrap_or(keyword_range.end()), source)?;
+
+        Ok(TextRange::new(keyword_range.start(), colon.end()))
+    }
+
+    /// Visits the nodes in the case header.
+    pub(crate) fn visit<F>(self, visitor: &mut F)
+    where
+        F: FnMut(AnyNodeRef),
+    {
+        fn visit<'a, N, F>(node: N, visitor: &mut F)
+        where
+            N: Into<AnyNodeRef<'a>>,
+            F: FnMut(AnyNodeRef<'a>),
+        {
+            visitor(node.into());
+        }
+
+        match self {
+            ClauseHeader::Class(class) => {
```

Comment:
Should we destructure these (and enumerate the fields) so that we're forced to update them if we modify the AST nodes?